src/main/kotlin/de/amplimind/codingchallenge/service/UserService.kt

```diff
@@ -121,38 +127,42 @@ class UserService(
      * @return the [UserInfoDTO] of the changed user
      */
     fun changeUserRole(changeUserRoleRequestDTO: ChangeUserRoleRequestDTO): UserInfoDTO {
-        if (changeUserRoleRequestDTO.newRole.matchesAny(UserRole.INIT)) {
-            // Cannot change user role to INIT
-            throw IllegalArgumentException("Cannot change user role to INIT")
-        }
+        try {
+            if (changeUserRoleRequestDTO.newRole.matchesAny(UserRole.INIT)) {
+                // Cannot change user role to INIT
+                throw IllegalArgumentException("Cannot change user role to INIT")
+            }
 
-        val user = UserUtils.fetchLoggedInUser()
+            val user = UserUtils.fetchLoggedInUser()
 
-        if (user.username == changeUserRoleRequestDTO.email) {
-            throw IllegalArgumentException("Cannot change own role")
-        }
+            if (user.username == changeUserRoleRequestDTO.email) {
+                throw IllegalArgumentException("Cannot change own role")
+            }
 
-        val foundUser =
-            this.userRepository.findByEmail(changeUserRoleRequestDTO.email)
-                ?: throw ResourceNotFoundException("User with email ${changeUserRoleRequestDTO.email} was not found")
+            val foundUser =
+                    this.userRepository.findByEmail(changeUserRoleRequestDTO.email)
+                            ?: throw ResourceNotFoundException("User with email ${changeUserRoleRequestDTO.email} was not found")
 
-        val updatedUser =
-            foundUser.let {
-                User(
-                    email = it.email,
-                    password = it.password,
-                    role = changeUserRoleRequestDTO.newRole,
-                )
-            }
+            val updatedUser =
+                    foundUser.let {
+                        User(
+                                email = it.email,
+                                password = it.password,
+                                role = changeUserRoleRequestDTO.newRole,
+                        )
+                    }
+            // save the updated user
+            this.userRepository.save(updatedUser)
 
-        // save the updated user
-        this.userRepository.save(updatedUser)
+            return UserInfoDTO(
+                    email = updatedUser.email,
+                    isAdmin = updatedUser.role.matchesAny(UserRole.ADMIN),
+                    status = extractUserStatus(updatedUser),
+            )
 
-        return UserInfoDTO(
-            email = updatedUser.email,
-            isAdmin = updatedUser.role.matchesAny(UserRole.ADMIN),
-            status = extractUserStatus(updatedUser),
-        )
+        } catch (ex: OptimisticLockingFailureException) {
+            throw IllegalStateException("The user was updated by another transaction")
```

Comment:
There is a retry mechanism in spring boot (where we can define a retry amount how often a certain action should be redone until a exception is thrown